gk/main.c

```diff
@@ -294,6 +310,7 @@ gk_process_request(struct flow_entry *fe, struct ipacket *packet,
 	} else {
```

Comment:
I cannot comment on the above ``if`` condition, i.e., ``if (priority < fe->u.request.last_priority && fe->u.request.allowance > 0)``. When  we have ``fe->u.request.allowance == 0``, it always goes to the ``else`` statement below regardless of the flow's waiting time, and refill the ``allowance`` field, then the attacker may be able to always send request traffic.

Reply:
The attacker always can send request traffic, the matter is the priority of that traffic. The allowance is meant to "hold" the same priority for a small number of packets. If an attacker keeps blasting packets, the allowance of a given priority will reach zero, and a new priority is going to be used (i.e. the `else` clause).

During an attack, the new priority is lower than the previous priority. To grasp what's going on, notice that at the beginning of `gk_process_request()` there is this update: `fe->u.request.last_packet_seen_at = now;` If the new priority is higher than the previous priority, it means that the attacker has backed off, that is, it did what we motivated it to do.